opennms-provision/opennms-provision-persistence/src/main/java/org/opennms/netmgt/provision/persist/DefaultForeignSourceService.java

```diff
@@ -259,30 +261,29 @@ public ForeignSource deletePolicy(String foreignSource, String name) {
      * @return a {@link java.util.Map} object.
      */
     @Override
-    public Map<String, String> getDetectorTypes() {
-        if (m_detectors == null) {
-            Map<String,String> detectors = new TreeMap<String,String>();
-            for (ServiceDetector d : m_serviceRegistry.findProviders(ServiceDetector.class)) {
-                String serviceName = d.getServiceName();
-                if (serviceName == null) {
-                    serviceName = d.getClass().getSimpleName();
-                }
-                String className = d.getClass().getName();
-                // NMS-8119: The class name may be changed when using proxy objects
-                if (d instanceof TargetClassAware) {
-                    className = ((TargetClassAware)d).getTargetClass().getName();
+    public Map<String, Class<?>> getDetectorTypes() {
+        if(m_detectors == null) {
+            m_detectors = new LinkedHashMap<>();
+        }
+        ServiceDetectorRegistry registry = m_serviceRegistry.findProvider(ServiceDetectorRegistry.class);
+        for(String serviceName: registry.getServiceNames()) {
+            if(!m_detectors.containsKey(serviceName)) {
+                CompletableFuture<ServiceDetectorFactory<?>> factoryFuture = registry.getDetectorClassNameFutureFromServiceName(serviceName)
+                        .thenCompose(registry::getDetectorFactoryFutureByClassName);
+                factoryFuture.whenComplete((f, e) ->m_detectors.put(serviceName, f.getDetectorClass()));
+                if (!factoryFuture.isDone()) {
+                    LOG.warn("The detector {} class not available.", serviceName);
```

Comment:
Can we add more details to log message that explain the behavior? It's not currently available, but what do we expect to happen?

Reply:
sure